File: test/ccall/entry.c
#include <stdio.h>

typedef struct {
    double lat;
    double lon;
} GeoCoord;

typedef struct {
    int numVerts;
    GeoCoord *verts;
} Geofence;

typedef struct {
    Geofence geofence;
    int numHoles;
    Geofence *holes;
} GeoPolygon;

int entry(Geofence* fence) {
    for (int idx =0; idx < fence->numVerts; idx++) {
      GeoCoord vert = fence->verts[idx];
      printf("vert %d %f %f\n", idx+1, vert.lat, vert.lon);
    }
    fflush(stdout);
    return 1;
}

int entry_polygon(GeoPolygon* polygon) {
    Geofence* fence = &(polygon->geofence);
    return entry(fence);
}

File: test/ccall/entry.jl
library_name = "libentry.A.dylib"
entry_c = normpath(@__DIR__, "entry.c")
const libentry = normpath(@__DIR__, library_name)
if !isfile(libentry)
    run(`clang -dynamiclib -std=c99 $entry_c -o $libentry`)
end

struct GeoCoord
    lat::Cdouble
    lon::Cdouble
end

struct Geofence
    numVerts::Cint
    verts::Ptr{GeoCoord}
end

struct GeoPolygon
    geofence::Geofence
    numHoles::Cint
    holes::Ptr{Geofence}
end

verts = [GeoCoord(0.01, 0.01),
         GeoCoord(0.01, -0.01),
         GeoCoord(-0.01, -0.01),
         GeoCoord(-0.01, 0.01)]
fence = Geofence(4, pointer(verts))
ccall((:entry, libentry), Int32, (Ptr{Geofence},), Ref(fence))
println()

polygon = GeoPolygon(fence, 0, C_NULL)
ccall((:entry_polygon, libentry), Int32, (Ptr{GeoPolygon},), Ref(polygon))
